protocols/luci-proto-bonding/htdocs/luci-static/resources/protocol/bonding.js

```diff
@@ -0,0 +1,442 @@
+		return _('You must select at least one ARP IP target if ARP monitoring is selected!');
+	}
+
+	return true;
+}
+
+function validate_primary_interface(section, value) {
+
+	var primary_valid = 'false';
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('slaves', section);
+	var selected_slaves = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('primary', section);
+	var selected_primary = opt[0].formvalue(section);
+
+	if (selected_policy == 'active-backup' || selected_policy == 'balance-tlb' || selected_policy == 'balance-alb') {
+
+		for (var i = 0; i < selected_slaves.length; i++) {
+			if (selected_slaves[i] == selected_primary) {
+				primary_valid = 'true';
+			}
+		}
+
+		if (primary_valid == 'false') {
+			return _('You must select a primary interface which is included in selected slave interfaces!');
+		}
+	}
+
+	return true;
+}
+
+return network.registerProtocol('bonding', {
+	getI18n: function() {
+		return _('Link Aggregation (Channel Bonding)');
+	},
+
+	getIfname: function() {
+		return null;
+	},
+
+	getOpkgPackage: function() {
+		return 'bonding';
+	},
+
+	isFloating: function() {
+		return true;
+	},
+
+	isCreateable: function(ifname) {
+		return getSelectableSlaves(ifname).then(L.bind(function(devices) {
+			return devices.length == 0 ?  _('No more slaves available') : null;
+		}, this));
+
+		return _('No more slaves available');
+	},
+
+	isVirtual: function() {
+		return true;
+	},
+
+	getDevices: function() {
+		return null;
+	},
+
+	containsDevice: function(ifname) {
+		return (network.getIfnameOf(ifname) == this.getIfname());
+	},
+
+	renderFormOptions: function(s) {
+		var o;
+
+		o = s.taboption('general', form.Value, 'ipaddr',
+				_('IPv4 address'),
+				_('The local IPv4 address'));
+		o.datatype = 'ip4addr';
+		o.validate = validateEmpty;
+
+		o = s.taboption('general', form.Value, 'netmask',
+				_('IPv4 netmask'),
+				_('The local IPv4 netmask'));
+		o.datatype = 'ip4addr';
+		o.validate = validateEmpty;
+		o.value("255.255.255.0");
+		o.value("255.255.0.0");
+		o.value("255.0.0.0");
+
+		o = s.taboption('advanced', form.MultiValue, 'slaves',
+				_('Slave Interfaces'),
+				_('Specifies which slave interfaces should be attached to this bonding interface'));
+
+		o.load = function(section_id) {
+			return getSelectableSlaves(section_id).then(L.bind(function(devices) {
+				for (var i = 0; i < devices.length; i++) {
+					this.value(devices[i], devices[i]);
+				}
+
+				if (devices.length == 0) {
+					this.placeholder = _('No more slaves available, can not save interface');
+					this.value('', '');
+					return '';
+				}
+
+				var if_slaves = uci.get('network', section_id, 'slaves');
+
+				if (if_slaves != null) {
+					return if_slaves;
+				}
+
+				return;
+			}, this));
+		};
+		o.cfgvalue = function(section_id) {
```

Comment:
This looks like the default implementation and can be dropped.

Reply:
OK, will do

Reply:
I dropped it according your suggestion.